File: robot/cpp/subsystems/VoiceControl.cpp
#include "subsystems/VoiceControl.h"

#include <algorithm>
#include <cctype>
#include <cstdio>
#include <iostream>
#include <memory>
#include <set>
#include <utility>

#include <frc/DriverStation.h>
#include <frc/RobotController.h>
#include <frc/smartdashboard/SmartDashboard.h>
#include <frc2/command/CommandScheduler.h>
#include <networktables/NetworkTableInstance.h>

namespace {
constexpr double kMaxAgeSeconds = 1.0;  // ignore stale/delayed commands
constexpr bool kAllowOnFms = false;     // ignore voice on a real field
const std::set<std::string> kAlwaysAllowed{"STOP"};  // works even when disabled

std::string Upper(std::string s) {
  std::transform(s.begin(), s.end(), s.begin(),
                 [](unsigned char c) { return std::toupper(c); });
  return s;
}
}  // namespace

VoiceControl::VoiceControl() {
  auto table = nt::NetworkTableInstance::GetDefault().GetTable("VoiceControl");
  m_commandSub = table->GetStringTopic("command").Subscribe("");
  m_armedSub = table->GetBooleanTopic("armed").Subscribe(false);

  // Whatever is already on the topic at startup is old news: never run it.
  if (auto initial = Parse(m_commandSub.Get())) {
    m_lastSeq = initial->seq;
  }

  Bind("STOP", [] { frc2::CommandScheduler::GetInstance().CancelAll(); });
  Log(2, "ready (verbosity " + std::to_string(m_verbosity) + ")");
}

void VoiceControl::Bind(std::string name, std::function<void()> action) {
  name = Upper(std::move(name));
  Log(2, "bound " + name);
  m_actions[name] = std::move(action);
}

void VoiceControl::BindCommand(std::string name, frc2::CommandPtr command) {
  auto owned = std::make_shared<frc2::CommandPtr>(std::move(command));
  Bind(std::move(name),
       [owned] { frc2::CommandScheduler::GetInstance().Schedule(*owned); });
}

void VoiceControl::Log(int level, const std::string& message) const {
  if (m_verbosity >= level) {
    std::cout << "[CMD] " << message << std::endl;
  }
}

void VoiceControl::Report(const std::string& name,
                          const std::string& status) const {
  frc::SmartDashboard::PutString("Voice/Last", name + ": " + status);
  Log(1, name + " " + status);
}

std::optional<VoiceControl::Parsed> VoiceControl::Parse(const std::string& raw) {
  const auto colon = raw.find(':');
  if (colon == std::string::npos || colon + 1 >= raw.size()) {
    return std::nullopt;
  }
  try {
    return Parsed{std::stoll(raw.substr(0, colon)), Upper(raw.substr(colon + 1))};
  } catch (const std::exception&) {
    return std::nullopt;
  }
}

bool VoiceControl::IsAllowed(const std::string& name) {
  if (kAlwaysAllowed.count(name) > 0) {
    return true;
  }
  if (frc::DriverStation::IsFMSAttached() && !kAllowOnFms) {
    return false;
  }
  return frc::DriverStation::IsEnabled();
}

void VoiceControl::Periodic() {
  frc::SmartDashboard::PutBoolean("Voice/Armed", m_armedSub.Get());

  auto atomic = m_commandSub.GetAtomic();
  auto parsed = Parse(atomic.value);
  if (!parsed || parsed->seq == m_lastSeq) {
    return;
  }
  m_lastSeq = parsed->seq;

  // NT server time and FPGA time share a clock on the roboRIO.
  const double ageSeconds =
      static_cast<double>(
          static_cast<int64_t>(frc::RobotController::GetFPGATime()) -
          atomic.serverTime) /
      1e6;

  char buf[96];
  std::snprintf(buf, sizeof buf, "rx seq=%lld name=%s age=%.2fs",
                static_cast<long long>(parsed->seq), parsed->name.c_str(),
                ageSeconds);
  Log(2, buf);

  auto it = m_actions.find(parsed->name);
  if (ageSeconds > kMaxAgeSeconds) {
    std::snprintf(buf, sizeof buf, "stale (%.1fs)", ageSeconds);
    Report(parsed->name, buf);
  } else if (it == m_actions.end()) {
    Report(parsed->name, "unbound");
  } else if (!IsAllowed(parsed->name)) {
    Report(parsed->name, "blocked (disabled/FMS)");
  } else {
    it->second();
    Report(parsed->name, "executed");
  }
}

File: robot/java/subsystems/VoiceControl.java
package frc.robot.subsystems;

import edu.wpi.first.networktables.BooleanSubscriber;
import edu.wpi.first.networktables.NetworkTable;
import edu.wpi.first.networktables.NetworkTableInstance;
import edu.wpi.first.networktables.StringSubscriber;
import edu.wpi.first.networktables.TimestampedString;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.RobotController;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.CommandScheduler;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

/**
 * Voice command receiver. Reads "<seq>:<COMMAND>" from NetworkTables table "VoiceControl" and runs
 * whatever you bound to COMMAND. Construct one in RobotContainer and call bind()/bindCommand().
 */
public class VoiceControl extends SubsystemBase {
  private static final String TABLE_NAME = "VoiceControl";
  private static final double MAX_AGE_SECONDS = 1.0; // ignore stale/delayed commands
  private static final boolean ALLOW_ON_FMS = false; // ignore voice on a real field
  private static final Set<String> ALWAYS_ALLOWED = Set.of("STOP"); // works even when disabled

  /**
   * Console verbosity, printed as "[CMD] ...":
   *
   * <ul>
   *   <li>0 = silent
   *   <li>1 = one line per command outcome (executed / stale / unbound / blocked)
   *   <li>2 = also log every received value and every binding
   * </ul>
   */
  private static final int DEFAULT_VERBOSITY = 1;

  private record Parsed(long seq, String name) {}

  private final StringSubscriber commandSub;
  private final BooleanSubscriber armedSub;
  private final Map<String, Runnable> actions = new HashMap<>();
  private long lastSeq = Long.MIN_VALUE;
  private int verbosity = DEFAULT_VERBOSITY;

  public VoiceControl() {
    NetworkTable table = NetworkTableInstance.getDefault().getTable(TABLE_NAME);
    commandSub = table.getStringTopic("command").subscribe("");
    armedSub = table.getBooleanTopic("armed").subscribe(false);

    // Whatever is already on the topic at startup is old news: never run it.
    Parsed initial = parse(commandSub.get());
    if (initial != null) {
      lastSeq = initial.seq();
    }

    bind("STOP", () -> CommandScheduler.getInstance().cancelAll());
    log(2, "ready (verbosity " + verbosity + ")");
  }

  public void setVerbosity(int level) {
    verbosity = level;
  }

  /** Run {@code action} immediately when {@code name} is heard. */
  public void bind(String name, Runnable action) {
    actions.put(name.toUpperCase(), action);
    log(2, "bound " + name.toUpperCase());
  }

  /** Schedule {@code command} when {@code name} is heard. */
  public void bindCommand(String name, Command command) {
    bind(name, () -> CommandScheduler.getInstance().schedule(command));
  }

  private void log(int level, String message) {
    if (verbosity >= level) {
      System.out.println("[CMD] " + message);
    }
  }

  private void report(String name, String status) {
    SmartDashboard.putString("Voice/Last", name + ": " + status);
    log(1, name + " " + status);
  }

  private static Parsed parse(String raw) {
    int colon = raw.indexOf(':');
    if (colon < 0 || colon == raw.length() - 1) {
      return null;
    }
    try {
      return new Parsed(
          Long.parseLong(raw.substring(0, colon)), raw.substring(colon + 1).trim().toUpperCase());
    } catch (NumberFormatException e) {
      return null;
    }
  }

  private static boolean isAllowed(String name) {
    if (ALWAYS_ALLOWED.contains(name)) {
      return true;
    }
    if (DriverStation.isFMSAttached() && !ALLOW_ON_FMS) {
      return false;
    }
    return DriverStation.isEnabled();
  }

  @Override
  public void periodic() {
    SmartDashboard.putBoolean("Voice/Armed", armedSub.get());

    TimestampedString atomic = commandSub.getAtomic();
    Parsed parsed = parse(atomic.value);
    if (parsed == null || parsed.seq() == lastSeq) {
      return;
    }
    lastSeq = parsed.seq();

    // NT server time and FPGA time share a clock on the roboRIO.
    double ageSeconds = (RobotController.getFPGATime() - atomic.serverTime) / 1e6;
    log(2, String.format("rx seq=%d name=%s age=%.2fs", parsed.seq(), parsed.name(), ageSeconds));

    Runnable action = actions.get(parsed.name());
    if (ageSeconds > MAX_AGE_SECONDS) {
      report(parsed.name(), String.format("stale (%.1fs)", ageSeconds));
    } else if (action == null) {
      report(parsed.name(), "unbound");
    } else if (!isAllowed(parsed.name())) {
      report(parsed.name(), "blocked (disabled/FMS)");
    } else {
      action.run();
      report(parsed.name(), "executed");
    }
  }
}